### PR TITLE
Implemented ASYNC client for communicating with HTTP/REST services

### PR DESCRIPTION
**Pull request by iagaponenko**

### CODE REVIEW

File: src/replica/HttpAsyncClient.cc
Comment (andy-slac) [inline src/replica/HttpAsyncClient.cc:151]: So, this does not actually throw, contrary to what doxygen comment says.

File: src/replica/HttpAsyncClient.h
Comment (andy-slac) [inline src/replica/HttpAsyncClient.h:205]: `std:set` is a heavy structure, maybe you want to use `std::initializer_list`?

File: src/replica/HttpAsyncClient.cc
Comment (andy-slac) [inline src/replica/HttpAsyncClient.cc:148]: The code in both `case`s is identical, consider merging.

File: src/replica/HttpAsyncClient.h
Comment (andy-slac) [inline src/replica/HttpAsyncClient.h:216]: _parsUrl -> _parseUrl?

File: src/replica/HttpAsyncClient.cc
Comment (andy-slac) [inline src/replica/HttpAsyncClient.cc:198]: This should be made static, no reason to compile regex on every call.

More importantly, your regex is really bad (sorry). Does not boost have a method for parsing URLs? No one should be using regexes for URL parsing, if you prefer I can show you how to call Python from C++ to do that 🙂 

Or at least write a unit test for this method with 100 or more different and legal URLs.

File: src/replica/HttpAsyncClient.cc
Comment (andy-slac) [inline src/replica/HttpAsyncClient.cc:204]: This sort of message is not super-helpful, saying "you have a problem" does not really identify what the problem is.

File: src/replica/HttpAsyncClient.cc
Comment (andy-slac) [inline src/replica/HttpAsyncClient.cc:208]: This is not the lambda that I can approve, even if it's short.

File: src/replica/HttpAsyncClient.cc
Comment (andy-slac) [inline src/replica/HttpAsyncClient.cc:216]: Grrr. You should definitely switch to Python.

File: src/replica/HttpAsyncClientApp.cc
Comment (andy-slac) [inline src/replica/HttpAsyncClientApp.cc:114]: Is this actually used?

File: src/replica/HttpAsyncClientApp.cc
Comment (andy-slac) [inline src/replica/HttpAsyncClientApp.cc:135]: Command line parser should do this check.

File: src/replica/HttpAsyncClientApp.cc
Comment (andy-slac) [inline src/replica/HttpAsyncClientApp.cc:185]: This is a very convoluted and inefficient way to write

    if (ptr->state() == HttpAsyncClient::State::FINISHED or ptr->state() == HttpAsyncClient::State::BODY_LIMIT_ERROR) {

**Review by andy-slac — APPROVED** [review]
I'm going to approve this, but I have to say that I have no confidence in the code that I saw. The logic in the async state machine looks unverifiable, and there is no unit test for it (and it's probably impossible to write a unit test for it). I would hate to debug this code when it breaks.

File: src/replica/HttpAsyncClient.h
Comment (iagaponenko) [reply src/replica/HttpAsyncClient.h:95]: This often happens (I have seen this many times) if there is silent memory corruption. Tracking the out-of-bound input values is very helpful in the early detection of problems.

File: src/replica/HttpAsyncClient.h
Comment (iagaponenko) [reply src/replica/HttpAsyncClient.h:112]: I need to parse URL by myself since we don't have a ready-to-use parser available in the current development container. There is an excellent Boost library that I would like to use https://master.url.cpp.al/. Though, it's not a part of the official Boost releases. The library is relatively new, and it requires some recent version of Boost that we don't have here. Ours is 1.66 (that was released circa 2017). The very same library could be used for parsing database connection strings. @fritzm is aware of the library.

File: src/replica/HttpAsyncClient.cc
Comment (iagaponenko) [reply src/replica/HttpAsyncClient.cc:208]: What's wrong with this lambda? Look at the next 3 lines of code where it's used. A value of the `in` parameter is extracted from the result of matching a regular expression. Using the lambda here simplifies the value analysis and prevets dereferencing the input value from the above mentioned result twice.

File: src/replica/HttpAsyncClient.cc
Comment (iagaponenko) [reply src/replica/HttpAsyncClient.cc:216]: Well, `11` is what the `Boost.Beast` expects from me. I wish we had (in C++) a good library that would be as convenient as Python's `requests` (AKA "HTTP for humans"). In theory, `Boost.Beast` provides a good fundation for building such library. However, this was not my intent here as developing such universally useful high-level library would take a much big effort.

File: src/replica/HttpAsyncClientApp.cc
Comment (iagaponenko) [reply src/replica/HttpAsyncClientApp.cc:114]: It was _meant_ to be used. I missed implementing a small piece of code that would do it. Thank you for noticing!

File: src/replica/HttpAsyncClientApp.cc
Comment (iagaponenko) [reply src/replica/HttpAsyncClientApp.cc:135]: You're 100% right. The command-line parser supports this. I will fix this.

File: src/replica/HttpAsyncClientApp.cc
Comment (iagaponenko) [reply src/replica/HttpAsyncClientApp.cc:185]: May be.

File: src/replica/HttpAsyncClient.h
Comment (iagaponenko) [reply src/replica/HttpAsyncClient.h:52]: Good point. I will fix this.

File: src/replica/HttpAsyncClient.h
Comment (iagaponenko) [reply src/replica/HttpAsyncClient.h:79]: This is fine with me. I will rename the class.

File: src/replica/HttpAsyncClient.h
Comment (iagaponenko) [reply src/replica/HttpAsyncClient.h:112]: Upon furthe rinspection of the code, I have realized that I already have the class `replica::Url` that could be extended to suit needs of the async client. Class `replica::Url` alredy has a unit test. I will update the code to use `replica::Url` before we have a better solution, like the previously mentioned `Boost.Url`.

File: src/replica/HttpAsyncClient.h
Comment (iagaponenko) [reply src/replica/HttpAsyncClient.h:151]: Let me explain the logic here. If a problem occurs in the constructor of the class (or its factory method `create`) then no object would be created. Once created, the object goes into the initial state `CREATED`. Both scenarios have no ambiguities. Things are getting a bit more complicated when the method `start` is attempting to initiate the request by setting up parameters of the request, initiating the expiration timer, and initiating resolve on the host's host/port. A failure to initiate the request will not automatically destroy the object (and all shared pointers to it). An exception is thrown to indicate a problem. At this point, it's not quite clear what should be done next and how to treat the object. The first possibility would be to leave the request in the state `CREATED` and assume that another call to the method `start` would "magically" resolve the issue. That's very unlikely because there is no way to "repair" the parameters of the object once it's constructed. Besides, it wouldn't be clear what exactly needs to be rapaired. The second approach that I chose here is to turn the object into the final state `FAILED` ... which also leaves the request object in an unusable state. On the surface, both are nearly the same. The only difference between those is that in the last approach the failed attempt to use the request object is now recorded in the object's state. Therefore, if this object was a part of some collection of requests, or if it was passed to some other context then the other context would be able to determine the new state of the object and take appropriate actions.

File: src/replica/HttpAsyncClient.h
Comment (iagaponenko) [reply src/replica/HttpAsyncClient.h:159]: Agreed

File: src/replica/HttpAsyncClient.h
Comment (iagaponenko) [reply src/replica/HttpAsyncClient.h:163]: The later. The HTTP code is only available once the response header was parsed. And this is only in the final states `FINISHED` or `BODY_LIMIT_ERROR`. In those cases the `HTTP` could be obtained by calling the method:
```
    /// @return The completion code (HTTP).
    /// @throw std::logic_error If the operation didn't start or it was still in progress.
    /// @note The header is available in states State::FINISHED and State::BODY_LIMIT_ERROR.
    int responseCode() const;
```

File: src/replica/HttpAsyncClient.h
Comment (iagaponenko) [reply src/replica/HttpAsyncClient.h:88]: The former. The request gets into the `FINISHED` state if both the response header and the body were received and parsed. It's up to a caller to inspect the response code by calling:
```
    /// @return The completion code (HTTP).
    /// @throw std::logic_error If the operation didn't start or it was still in progress.
    /// @note The header is available in states State::FINISHED and State::BODY_LIMIT_ERROR.
    int responseCode() const;
```
I should perhaps make this more clear in the documentation for the class.

File: src/replica/HttpAsyncClient.h
Comment (iagaponenko) [reply src/replica/HttpAsyncClient.h:169]: My bad.

File: src/replica/HttpAsyncClient.h
Comment (iagaponenko) [reply src/replica/HttpAsyncClient.h:175]: Very good suggestion! I missed to address this complication.

File: src/replica/HttpAsyncClient.cc
Comment (iagaponenko) [reply src/replica/HttpAsyncClient.cc:75]: I have already explained the motivation for catching the out-of-band values in another comment posted before.

File: src/replica/HttpAsyncClient.cc
Comment (iagaponenko) [reply src/replica/HttpAsyncClient.cc:126]: I agree regarding the first two methods. It does make sense to move both to the c-tor. The third method should be initiated here because once created, the request object might be stored in some queue and be initiated at the right time (once it's needed). That was the conscious decision to separate object construction from any network operations. Hence its has two states in its FSA:
```
CREATE -> IN_PROGRESS -> 
```

File: src/replica/HttpAsyncClient.cc
Comment (iagaponenko) [reply src/replica/HttpAsyncClient.cc:129]: You're next to @fritzm who doesn't like `bind`. I still don't get why. What's wrong with those? Is this just your personal opinion, or advice from the C++ "gurus" (Alexandersku, Sutter, Meyer, etc.)? In any case, I would be curious to see the very source of this opinion. Then, perhaps, I will reconsider my "obsession" with `bind` (I use it quite a bit in other classes).

File: src/replica/HttpAsyncClient.cc
Comment (iagaponenko) [reply src/replica/HttpAsyncClient.cc:148]: Good point.

File: src/replica/HttpAsyncClient.cc
Comment (iagaponenko) [reply src/replica/HttpAsyncClient.cc:198]: This question was addressed in response to the relevant comment.

File: src/replica/HttpAsyncClient.cc
Comment (iagaponenko) [reply src/replica/HttpAsyncClient.cc:204]: Noted. I will improve the message.

**Comment by iagaponenko** [comment]
@andy-slac regarding your general comment on the unit test, I think it's possible to implement the one here. I could launch the in-process `qhttp` on some free port and use this server for testing the new class.

File: src/replica/HttpAsyncClient.h
Comment (andy-slac) [reply src/replica/HttpAsyncClient.h:95]: That's an odd argument. If you have a memory corruption I think you cannot even reason about what your program does.

File: src/replica/HttpAsyncClient.h
Comment (andy-slac) [reply src/replica/HttpAsyncClient.h:151]: This is fine, my question was mostly that it's not clear what can cause this method to fail, as I thought that most checks and hostname/port parsing/resolution should be done in constructor.

File: src/replica/HttpAsyncClient.h
Comment (iagaponenko) [reply src/replica/HttpAsyncClient.h:95]: At least, I will see the problem earlier. This helped in the past. The out-of-bound enumerator values retrieved from the data members of some class usually indicate that an object of that class got destroyed, corrupt, or never created. This is never a problem in Python because Python doesn't know anything about such things as "threads" or "event-driven" programming models 😄 
What I could do here is to remove the notion of the exception from the class contract while still leaving this extra check in the body of the method. This may prevent confusion.

File: src/replica/HttpAsyncClient.h
Comment (andy-slac) [reply src/replica/HttpAsyncClient.h:163]: Maybe change "error" to "error message"?

File: src/replica/HttpAsyncClient.cc
Comment (andy-slac) [reply src/replica/HttpAsyncClient.cc:75]: Frankly, that comment does not make sense.

File: src/replica/HttpAsyncClient.h
Comment (iagaponenko) [reply src/replica/HttpAsyncClient.h:151]: I will fix the documentation on the method along with pushing parameter validation to the c-tor.

File: src/replica/HttpAsyncClient.h
Comment (iagaponenko) [reply src/replica/HttpAsyncClient.h:163]: Good suggestion!

File: src/replica/HttpAsyncClient.cc
Comment (andy-slac) [reply src/replica/HttpAsyncClient.cc:129]: We all C++ gurus, no need to call other authorities 🙂  Bind sucks because it makes code unreadable and its implementation is just terrible.

File: src/replica/HttpAsyncClient.cc
Comment (andy-slac) [reply src/replica/HttpAsyncClient.cc:208]: Couple of points:
- I don't see this as an improvement, it makes code longer but not clearer
- the name is confusing, `parseVal` does not parse anything, just assigns something to something
- if you care about performance you should probably write something like `if (match[1].matched) _host = match[1];` to avoid creating strings to just check that they are empty.

File: src/replica/HttpAsyncClient.cc
Comment (fritzm) [reply src/replica/HttpAsyncClient.cc:129]: Yup, strongly agree w/ Andy on this one, per previous ~arguments~ "discussions".  I don't understand why you embrace modern C++ in so many other ways, but still cling so stubbornly to `bind`?

File: src/replica/HttpAsyncClient.cc
Comment (fritzm) [reply src/replica/HttpAsyncClient.cc:151]: Doxygen tag-line: "ugly, useless, and frequently a pack of lies"

**Review by fritzm — CHANGES_REQUESTED** [review]
Regarding the unit tests:

I still don't see a real need for the AsyncTimer class here, and consider it a dangerous addition to our codebase because it contains an asio anti-pattern (proliferation of io_service instances in class-private threads) and has a high potential for misuse.

At the very least, before merging I would want to see the timer class moved into the unit test cpp/h anonymous namespace instead of separate files.

Much better in my opinion would be to just use a single `io_service` in each test (shared for server, client, and timeout timer) and use the existing asio timer API out of the box, without any need for an additional wrapper class at all.  We did discuss some of this previously, but after seeing the tests in their final form I'm less convinced than previously that the new AsyncTimer class is well-justified.

File: src/replica/testHttpAsyncReq.cc
Comment (fritzm) [inline src/replica/testHttpAsyncReq.cc:202]: Do you really want `exit()` here?  If one test fails, shouldn't the others still be attempted?

File: src/replica/testHttpAsyncReq.cc
Comment (fritzm) [inline src/replica/testHttpAsyncReq.cc:336]: This `sleep` is probably jamming the server (which has one thread) from processing further requests (and if you had run the entire test on a single thread would jam the entire test).  Since you are in an asio thread here, instead of sleep, it would probably be better to use an async timer (capture the response pointer in the completion lambda)

File: src/replica/testHttpAsyncReq.cc
Comment (fritzm) [inline src/replica/testHttpAsyncReq.cc:492]: Fixed port is not great, but probably okay.  You might want to pick something less commonly used than 8080?

An alternative might be to create a new asio socket here and `bind()` it to an address with port 0.  This will allocate an unused port.  `qhttp` uses SO_REUSEADDR option already on its listen port, so you could leave the socket open and just pass that newly allocated port to the `qhttp::Server` constructor, and I think it would work and avoid the potential collision problem?

**Review by fritzm — APPROVED** [review]
Okay, thanks for changes per previous discussion.  A few additional comments left for your consideration.

File: src/replica/testHttpAsyncReq.cc
Comment (iagaponenko) [reply src/replica/testHttpAsyncReq.cc:202]: I too thought about this. I will see how to improve this.

File: src/replica/testHttpAsyncReq.cc
Comment (iagaponenko) [reply src/replica/testHttpAsyncReq.cc:336]: Yes, I'm aware of this risk. Fortunatelly, this is not a problem here because `qhttp` has its own instance of `io_service` and a dedicated thread. Besides, each of the test cases initiates no more than one request.

File: src/replica/testHttpAsyncReq.cc
Comment (iagaponenko) [reply src/replica/testHttpAsyncReq.cc:492]: I will pick some other (large number) port.

As far as your `bind` idea is concerned, I have already explored a kind of similar track by starting and stopping `qhttp` using this series of calls:
```
auto ptr = qhttp::create(io_service, 0);
ptr->start();
ptr->stop();
...
// Then delayed start in the ASYNC timer callback hoping the server would get the same port
{
  ptr->start()
}
```
This all ended up with some really bad crash in ... `XRootD` stack 😸 

I could also modify this algorithm by dropping the first instance of `qhttp` and replacing it with another one. Would give me the same result as your `bind` proposal?

**Pull request merged**
